File: Two water Jug problem.java
class Solution {
    public int minSteps(int m, int n, int d) {
        // code here
        if (d > Math.max(m, n) || d % gcd(m, n) != 0){
            return -1;
        }
        if (d == 0){
            return 0;
        }
        return Math.min(pour(m, n, d), pour(n,m,d));
    }
    private int pour(int a, int b, int d){
        int count = 0;
        int x = 0, y = 0;
        
        while (x != d && y != d){
            if (x == 0){
                x = a;
                count++;
            } else if (y == b){
                y = 0;
                count++;
            } else {
                int pour_a = Math.min(x, b - y);
                x -= pour_a;
                y += pour_a;
                count++;
            }
        }
        return count;
    }
    private int gcd(int a, int b){
        while (b != 0){
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}

File: Two water Jug problem.py
from math import gcd
class Solution:
	def minSteps(self, m, n, d):
		# Code here
		if d > max(m, n) or d % gcd(m, n) != 0:
		    return -1
		if d == 0:
		    return 0
		def pour(a, b):
		    count = 0
		    x, y = 0, 0
		    while x != d and y != d:
		        if x == 0:
		            x = a
		            count += 1
		        elif y == b:
		            y = 0
		            count += 1
		        else:
		            pour_a = min(x, b - y)
		            x -= pour_a
		            y += pour_a
		            count += 1
		    return count
        return min(pour(m, n), pour(n, m))                count++;

